### DISCUSSION TITLE
zh/modes/track/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8563] [Docs]
# zh/modes/track/

了解如何使用Ultralytics YOLO  在视频流中进行物体跟踪。使用不同跟踪器和自定义跟踪器配置的指南。

https://docs.ultralytics.com/zh/modes/track/

**Comment by Eric5619** [comment]


**Reply by glenn-jocher** [reply]
Hey there! 😊 To access the tracking ID of each object when using Ultralytics YOLO for object tracking, you can simply refer to the `id` attribute of the `boxes` object in your results. Here's a quick example using Python:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('yolov8n.pt')  # or your custom model

# Perform tracking
results = model.track(source='path/to/your/video.mp4')

# Access tracking IDs
for result in results:
    track_ids = result.boxes.id
    print(track_ids)
```

This will print out the tracking IDs for the objects detected in each frame of your video. Hope this helps! Let me know if you have any more questions.

**Comment by YongerZhu** [comment]
Can I use YOLO V8 to track the same object on different surveillance screens, and if so, how?

**Reply by pderrenger** [reply]
Absolutely! You can use YOLOv8 for tracking objects across different surveillance screens. Here's a quick guide on how to do it:

1. **Load Your Model**: First, load your YOLOv8 model. You can use a pre-trained model or your custom-trained model.
   ```python
   from ultralytics import YOLO
   model = YOLO('yolov8n.pt')  # or your custom model
   ```

2. **Run the Tracker**: Use the `track` method on your model and specify the source as your video or stream URL. You can choose from different trackers like BoT-SORT or ByteTrack by specifying the `tracker` argument.
   ```python
   results = model.track(source="path/to/your/video.mp4", tracker="bytetrack.yaml")
   ```

3. **Process Results**: The `results` object will contain the tracked objects with unique IDs that persist across frames.

Remember, for tracking across different screens, you'll need to run the tracker separately for each video source and then correlate the tracking IDs based on your specific requirements, such as appearance similarity or spatial proximity.

Happy tracking! 🚀

**Reply by YongerZhu** [reply]
Thank you for your reply. What I would like to know further is that if a vehicle passes through a stretch of road, and there are multiple cameras along this road, and each camera is able to capture the vehicle's movement within its field of view separately, how do I match correlate that vehicle from different cameras?

Can I do this with just YOLO V8, or do I need to use some feature matching algorithm to associate the same vehicle from different cameras based on its appearance?

**Reply by pderrenger** [reply]
@YongerZhu hey there! 🌟 For tracking the same vehicle across multiple cameras, YOLOv8 can detect and track the vehicle within each camera's view. However, to correlate the vehicle across different cameras, you'd typically need an additional step involving a feature matching algorithm or a multi-camera tracking system. This step would compare features or tracking IDs to associate the same vehicle across different camera feeds based on its appearance, trajectory, or other distinguishing features.

While YOLOv8 excels in detection and tracking within individual video streams, integrating these detections across multiple streams for consistent tracking of an object involves more complex logic and potentially additional tools or algorithms focused on multi-camera tracking scenarios. 🚗💨

For a start, you could use YOLOv8 to extract features or bounding boxes and then apply algorithms like cosine similarity for feature matching or use spatial and temporal information for cross-camera correlation. Happy coding!

**Reply by YongerZhu** [reply]
thank you for your help

**Reply by pderrenger** [reply]
Hey there! 🌟 Thanks for reaching out. To track a vehicle across multiple cameras, you'll indeed need a bit more than just YOLOv8's detection and tracking capabilities. After using YOLOv8 to detect and track the vehicle within each camera's view, you'd typically employ a feature matching algorithm or a multi-camera tracking system to correlate the vehicle's appearance across different cameras. This could involve comparing features like color histograms, shape, or even using more advanced techniques like deep learning-based re-identification (Re-ID) models that can generate a unique fingerprint for each vehicle. 

Here's a simplified example of how you might approach this with YOLOv8 detections:

```python
# Assume you have detections from YOLOv8 for each camera
detections_cam1 = yolo_model.detect(cam1_frame)
detections_cam2 = yolo_model.detect(cam2_frame)

# Extract features for each detection (simplified)
features_cam1 = extract_features(detections_cam1)
features_cam2 = extract_features(detections_cam2)

# Match features across cameras
matches = match_features(features_cam1, features_cam2)
```

This is a high-level overview, and the actual implementation can get quite complex, especially in handling varying angles, lighting conditions, and occlusions across cameras. Tools like OpenCV can be helpful for feature extraction and matching, and there are several research papers and projects on multi-camera tracking and vehicle re-identification that might offer deeper insights and methodologies. Happy coding! 🚀

**Comment by Fancyhzw** [comment]
How can I use YOLO to show the center point in the window?

**Reply by glenn-jocher** [reply]
Hey there! 😊 To show the center point of detected objects using YOLO, you can modify the plotting function to include the center points. Here's a quick example assuming you're using the Python API:

```python
from ultralytics import YOLO
import cv2

# Load your model
model = YOLO('yolov8n.pt')

# Run inference
results = model('path/to/your/image.jpg')

# Extract boxes
boxes = results[0].boxes.xyxy.cpu().numpy()

# Calculate and plot center points
for box in boxes:
    center_x = int((box[0] + box[2]) / 2)
    center_y = int((box[1] + box[3]) / 2)
    cv2.circle(results[0].orig_img, (center_x, center_y), radius=5, color=(0, 255, 0), thickness=-1)

# Display the image with center points
cv2.imshow('YOLO Centers', results[0].orig_img)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

This code snippet will draw a green dot at the center of each detected object in your image. Hope this helps! Let me know if you have any other questions.

**Comment by Daniel-lucifer** [comment]
After accessing the tracking ID of each object when using Ultratics YOLO for object tracking, if I use several images I have collected myself, such as three images, how can I match the features detected in the three images I provide? Because some features may appear repeatedly

**Reply by pderrenger** [reply]
Hello! It's great to see you exploring object tracking with YOLO and using custom datasets. 😊 When it comes to matching detected features across multiple images, especially where the same features might appear in each, there are a few strategies you might consider:

1. **Feature Matching Approaches:** You can utilize feature matching algorithms like SIFT (Scale-Invariant Feature Transform) or ORB (Oriented FAST and Rotated BRIEF) to find similarities between features extracted from your images. Here’s a simple snippet using ORB with OpenCV for matching features between two images:

```python
import cv2

# Load images
img1 = cv2.imread('image1.jpg', 0)  # Query image
img2 = cv2.imread('image2.jpg', 0)  # Train image

# Initialize ORB detector
orb = cv2.ORB_create()

# Find the keypoints and descriptors
kp1, des1 = orb.detectAndCompute(img1, None)
kp2, des2 = orb.detectAndCompute(img2, None)

# Create BFMatcher object and match descriptors
bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)
matches = bf.match(des1, des2)

# Sort them in ascending order of their distances so that best matches come to front
matches = sorted(matches, key=lambda x: x.distance)

# Draw first 10 matches.
img3 = cv2.drawMatches(img1, kp1, img2, kp2, matches[:10], None, flags=2)
cv2.imshow('Matches', img3)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

2. **Tracking Across Frames:** If your images are frames from a video or a sequence where objects move slightly between frames, you might employ tracking algorithms like the ones already integrated into YOLO's tracking features (e.g., BoT-SORT, ByteTrack) which you have seen in the Ultralytics documentation. These trackers inherently handle feature matching across frames to maintain consistent IDs for each object.

3. **Custom Solutions:** Depending on the specificity of your task (for example, matching unique objects across different scene conditions), training a neural network to recognize specific features or objects across your dataset might provide more tailored results.

Matching features reliably across multiple images can be challenging, especially in varied conditions. It's often a combination of techniques and some fine-tuning that leads to the best results.

Hope this helps! If you have further questions or need more detailed assistance, feel free to ask. Happy coding! 😄

**Comment by Daniel-lucifer** [comment]
How can I determine my position in this scene based on three images collected at a fixed location after accessing the tracking ID of each object when using Ultratics YOLO for object tracking

**Reply by glenn-jocher** [reply]
确定场景中的位置基于三张在固定位置收集的图像以及物体的跟踪ID时，可以通过比较物体的相对位置和运动轨迹在连续的帧中变化来实现。若您能够获取每个物体在这三张图像中的坐标或者位置信息，就可以利用这些数据来估计您相对于这些物体的位置变化。下面是一个简单的例子，说明如何使用Ultralytics YOLO进行物体跟踪，并获取每个追踪物体的坐标。

```python
from ultralytics import YOLO

# Initialize model
model = YOLO('path/to/your/model.pt') 

# Perform tracking
results = model.track("path/to/your/video.mp4")

for frame in results:
    boxes = frame.boxes.xyxy  # xyxy format
    track_ids = frame.boxes.id
    for track_id, box in zip(track_ids, boxes):
        x1, y1, x2, y2 = box
        print(f"Track ID: {track_id} is at position: [({x1}, {y1}), ({x2}, {y2})]")
```

这段代码运行后，您将能看到每一个追踪物体在每一帧中的位置。根据这些位置信息，您可以分析物体随时间的移动轨迹，进而推断出您的位置。希望这对您有所帮助！🚀

**Reply by Ares-cz** [reply]
If I want to use YOLO11s-seg.pt to conduct MOTS task, is the result in save_txt like this:

0 0.626562 0.608333 0.626562 0.6375 0.629687 0.6375 0.630469 0.638889 0.63125 0.638889 0.632031 0.640278 0.632812 0.640278 0.635938 0.634722 0.635938 0.633333 0.636719 0.631944 0.636719 0.629167 0.6375 0.627778 0.6375 0.626389 0.638281 0.625 0.638281 0.623611 0.641406 0.618056 0.641406 0.615278 0.642187 0.613889 0.644531 0.613889 0.644531 0.608333 1
0 0.323438 0.641667 0.323438 0.684722 0.326562 0.684722 0.328125 0.6875 0.328906 0.6875 0.329688 0.688889 0.334375 0.688889 0.335156 0.6875 0.338281 0.6875 0.338281 0.665278 0.335938 0.665278 0.335156 0.663889 0.335156 0.659722 0.334375 0.658333 0.334375 0.654167 0.333594 0.652778 0.333594 0.651389 0.332812 0.65 0.332812 0.641667 2

The last number is track_id? and can I use frame.masks.id? or frame.masks.xyxy?

In addition, are there any tools in ultralytics can transfer the x1 y1 x2 y2 ... xn, yn to RLE

Thank you very much!

**Comment by GKYoffice** [comment]
I want to get some performance indicators about trackers, such as MOTA, IDF1, etc. How should I do it?

**Reply by glenn-jocher** [reply]
To evaluate trackers using metrics such as MOTA, IDF1, etc., you would typically use specialized tracking evaluation tools that are capable of computing these metrics based on the tracker's output. 

At present, direct support for these evaluation metrics isn't built into the Ultralytics YOLO platform. However, you can integrate tracking output with evaluation frameworks like [MOTChallenge](https://motchallenge.net/) which provide tools to evaluate multiple object tracking performance using these metrics.

Here is a general approach on how you might proceed:
1. Export your tracking results in a format compatible with the evaluation tool (e.g., `.txt` or `.csv` with frame ID, track ID, bounding boxes, etc.).
2. Use the evaluation tool to read these results and the corresponding ground truth.
3. Compute the metrics.

For specific implementation, you might refer to the documentation or forums associated with the chosen evaluation tool for detailed steps and code examples. If integrating new evaluation measures directly into your workflows is of interest, consider contributing to the Ultralytics repository to include these features.

**Comment by zhaozhaotnl** [comment]
Hello! I am so glad to use yolov8,but I have a question that which tracking method the under code use.Bytetrack or Botsort?
from collections import defaultdict

import cv2
import numpy as np

from ultralytics import YOLO

# Load the YOLOv8 model
model = YOLO('yolov8n.pt')

# Open the video file
video_path = "path/to/video.mp4"
cap = cv2.VideoCapture(video_path)

# Store the track history
track_history = defaultdict(lambda: [])

# Loop through the video frames
while cap.isOpened():
    # Read a frame from the video
    success, frame = cap.read()

    if success:
        # Run YOLOv8 tracking on the frame, persisting tracks between frames
        results = model.track(frame, persist=True)

        # Get the boxes and track IDs
        boxes = results[0].boxes.xywh.cpu()
        track_ids = results[0].boxes.id.int().cpu().tolist()

        # Visualize the results on the frame
        annotated_frame = results[0].plot()

        # Plot the tracks
        for box, track_id in zip(boxes, track_ids):
            x, y, w, h = box
            track = track_history[track_id]
            track.append((float(x), float(y)))  # x, y center point
            if len(track) > 30:  # retain 90 tracks for 90 frames
                track.pop(0)

            # Draw the tracking lines
            points = np.hstack(track).astype(np.int32).reshape((-1, 1, 2))
            cv2.polylines(annotated_frame, [points], isClosed=False, color=(230, 230, 230), thickness=10)

        # Display the annotated frame
        cv2.imshow("YOLOv8 Tracking", annotated_frame)

        # Break the loop if 'q' is pressed
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        # Break the loop if the end of the video is reached
        break

# Release the video capture object and close the display window
cap.release()
cv2.destroyAllWindows()

**Reply by pderrenger** [reply]
Hello! I'm glad you're enjoying using YOLOv8! In your script, the specific tracking method used (ByteTrack or BoT-SORT) depends on the configuration you've set when initializing the `YOLO()` object or the default settings if none is specified explicitly.

By default, YOLOv8 uses BoT-SORT as the tracker unless overriden by specifying a different tracker configuration. If you want to explicitly use ByteTrack or any other tracker, you can set the `tracker` argument like this:

```python
from ultralytics import YOLO

# Load a YOLOv8 model with ByteTrack
model = YOLO('yolov8n.pt', tracker='bytetrack.yaml')

# Now proceed with your tracking code
```

Feel free to swap `'bytetrack.yaml'` with the specific tracker config file you are interested in using. Enjoy tracking! 🚀

**Reply by zhaozhaotnl** [reply]
Thanks for answering!It works. Thanks again!

**Reply by pderrenger** [reply]
I'm thrilled to hear that the solution worked for you! If you have any more questions or need further assistance as you continue exploring YOLOv8, feel free to reach out. Happy coding! 😊🚀

**Comment by XiaoMingWM** [comment]
Can YOLOV8 do object tracking on the image to distinguish between different ids？

**Reply by pderrenger** [reply]
@XiaoMingWM yes, YOLOv8 can perform object tracking and distinguish between different IDs. It supports multiple tracking algorithms like BoT-SORT and ByteTrack, which can be enabled using specific YAML configuration files. This allows the model to track objects across video frames and maintain unique IDs for each detected object, making it suitable for applications such as surveillance, sports analytics, and more.

For more details on how to implement and customize tracking with YOLOv8, you can refer to our comprehensive guide on multi-object tracking available in the documentation. If you have any more questions or need further assistance, feel free to ask! 🚀

**Comment by zhaozhaotnl** [comment]
from collections import defaultdict

import cv2
import numpy as np
from ultralytics import YOLO

# Load the YOLOv8 model
model = YOLO("yolov8n.pt",tracker='bytetrack.yaml')

# Open the video file
video_path = "path/to/video.mp4"
cap = cv2.VideoCapture(video_path)

# Store the track history
track_history = defaultdict(lambda: [])

# Loop through the video frames
while cap.isOpened():
    # Read a frame from the video
    success, frame = cap.read()

    if success:
        # Run YOLOv8 tracking on the frame, persisting tracks between frames
        results = model.track(frame, persist=True)

        # Get the boxes and track IDs
        boxes = results[0].boxes.xywh.cpu()
        track_ids = results[0].boxes.id.int().cpu().tolist()

        # Visualize the results on the frame
        annotated_frame = results[0].plot()

        # Plot the tracks
        for box, track_id in zip(boxes, track_ids):
            x, y, w, h = box
            track = track_history[track_id]
            track.append((float(x), float(y)))  # x, y center point
            if len(track) > 30:  # retain 90 tracks for 90 frames
                track.pop(0)

            # Draw the tracking lines
            points = np.hstack(track).astype(np.int32).reshape((-1, 1, 2))
            cv2.polylines(annotated_frame, [points], isClosed=False, color=(230, 230, 230), thickness=10)

        # Display the annotated frame
        cv2.imshow("YOLOv8 Tracking", annotated_frame)

        # Break the loop if 'q' is pressed
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        # Break the loop if the end of the video is reached
        break

# Release the video capture object and close the display window
cap.release()
cv2.destroyAllWindows()


The code is error.
Traceback (most recent call last):
  File "/root/ultralytics-main/yolo.py", line 8, in <module>
    model = YOLO("yolov8n.pt",tracker='bytetrack.yaml')
TypeError: __init__() got an unexpected keyword argument 'tracker'
can i get help?

**Reply by glenn-jocher** [reply]
Hello! It looks like you're trying to initialize the `YOLO` model with a `tracker` argument, which isn't supported directly in the constructor. Instead, you should load the model first and then specify the tracker when calling the `track` method. Here's how you can adjust your code:

```python
from ultralytics import YOLO

# Load the YOLOv8 model
model = YOLO("yolov8n.pt")

# Specify the tracker when calling the track method
results = model.track("path/to/video.mp4", tracker='bytetrack.yaml', persist=True)
```

This way, you load the model first and then apply the tracker during the tracking process. If you have any more questions or need further assistance, feel free to ask!

**Reply by zhaozhaotnl** [reply]
Thanks!I will try this.Thanks for your code!

**Reply by glenn-jocher** [reply]
You're welcome! I'm glad you found the code helpful. If you encounter any issues or need further assistance, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, providing a minimum reproducible example would be very helpful for us to investigate further. You can refer to our guide on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

For your tracking needs, remember to load the model first and then specify the tracker when calling the `track` method. Here's a refined version of your code:

```python
from collections import defaultdict
import cv2
import numpy as np
from ultralytics import YOLO

# Load the YOLOv8 model
model = YOLO("yolov8n.pt")

# Open the video file
video_path = "path/to/video.mp4"
cap = cv2.VideoCapture(video_path)

# Store the track history
track_history = defaultdict(lambda: [])

# Loop through the video frames
while cap.isOpened():
    # Read a frame from the video
    success, frame = cap.read()

    if success:
        # Run YOLOv8 tracking on the frame, persisting tracks between frames
        results = model.track(frame, persist=True, tracker='bytetrack.yaml')

        # Get the boxes and track IDs
        boxes = results[0].boxes.xywh.cpu()
        track_ids = results[0].boxes.id.int().cpu().tolist()

        # Visualize the results on the frame
        annotated_frame = results[0].plot()

        # Plot the tracks
        for box, track_id in zip(boxes, track_ids):
            x, y, w, h = box
            track = track_history[track_id]
            track.append((float(x), float(y)))  # x, y center point
            if len(track) > 30:  # retain 90 tracks for 90 frames
                track.pop(0)

            # Draw the tracking lines
            points = np.hstack(track).astype(np.int32).reshape((-1, 1, 2))
            cv2.polylines(annotated_frame, [points], isClosed=False, color=(230, 230, 230), thickness=10)

        # Display the annotated frame
        cv2.imshow("YOLOv8 Tracking", annotated_frame)

        # Break the loop if 'q' is pressed
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        # Break the loop if the end of the video is reached
        break

# Release the video capture object and close the display window
cap.release()
cv2.destroyAllWindows()
```

Feel free to reach out if you have any more questions. Happy coding! 🚀

For more detailed information on tracking, you can visit our [tracking documentation](https://docs.ultralytics.com/modes/track/).

**Comment by jianghaokun** [comment]
How can I deal with the object ID discontinuity when using yolov8 for multi-object tracking? For example, ID 6 is missing from ID 1 to ID 10.

**Reply by pderrenger** [reply]
@jianghaokun object ID discontinuity, such as missing IDs, can occur due to the nature of the tracking algorithms and the confidence thresholds set for object detection. To mitigate this, ensure that your `track_high_thresh` is appropriately set to avoid losing track of objects with lower confidence scores. Additionally, using a more robust tracker like BoT-SORT or ByteTrack can help maintain consistent IDs. If the issue persists, consider fine-tuning your model or tracker configurations. For further details, you can refer to the [tracking documentation](https://docs.ultralytics.com/modes/track/).

**Comment by zhaozhaotnl** [comment]
Hello.I have a question. If I need to modify the network structure of bytetrack, do I only need to modify some of the code in yolov8 or do I need to copy the code of bytetrack from github and combine yolov8 with bytetrack before modifying the code of bytetrack.please help me to solve this question.

**Reply by glenn-jocher** [reply]
@zhaozhaotnl to modify the network structure of ByteTrack within YOLOv8, you will need to integrate the ByteTrack code into the YOLOv8 framework. This involves copying the necessary ByteTrack code from its GitHub repository and combining it with YOLOv8. After integration, you can then modify the ByteTrack code as needed. Ensure you are using the latest versions of both YOLOv8 and ByteTrack to avoid compatibility issues. For more detailed guidance, you can refer to the Ultralytics documentation on tracking: https://docs.ultralytics.com/modes/track/.

**Reply by zhaozhaotnl** [reply]
Thanks for answering!I will try.Thanks again.

**Reply by glenn-jocher** [reply]
You're welcome! If you encounter any issues or need further assistance, feel free to ask. Good luck with your project! For more detailed information on tracking, you can refer to the Ultralytics documentation: https://docs.ultralytics.com/modes/track/.

**Comment by MarserHK** [comment]
Hello, I have a question. My detection model includes multiple categories. How can I specify to track only a certain category?

**Reply by glenn-jocher** [reply]
@MarserHK to track only a specific category, you can filter the detections by class before passing them to the tracker. Use the `classes` argument in the `track` method to specify the class IDs you want to track. For example, if you want to track only class ID 0, you can set `classes=[0]`. Here’s a brief example:

```python
results = model.track(source="your_video.mp4", classes=[0])
```

This will ensure that only objects of the specified class are tracked. For more details, you can refer to the [Ultralytics documentation on tracking](https://docs.ultralytics.com/modes/track/).

**Comment by 1852328953** [comment]
如何使用bytetrack跟踪算法

**Reply by glenn-jocher** [reply]
@1852328953 to use the ByteTrack algorithm with Ultralytics YOLO, load your model and specify the tracker configuration using `tracker="bytetrack.yaml"` in your tracking command. For more details, refer to the [tracking documentation](https://docs.ultralytics.com/modes/track/).

**Comment by xiao5676wen** [comment]
I think of MaxID, what do I need to do?

**Reply by glenn-jocher** [reply]
@xiao5676wen to address your question about "MaxID," you can specify the `max_id` parameter in the tracking configuration file (such as `botsort.yaml` or `bytetrack.yaml`). This parameter controls the maximum ID value for tracked objects. You can modify and use a custom tracker configuration file for your needs. For further details, visit [Ultralytics Trackers Documentation](https://docs.ultralytics.com/modes/track/#available-trackers). Let me know if you need additional assistance!

**Reply by xiao5676wen** [reply]
I want to get the maximum value of the ID displayed on the detection box in the video for each target tracking! Is that value all the quantities of the target? I just want to get that value!

**Reply by xiao5676wen** [reply]
Because I am doing single object detection, I know that the last ID value is the result of my previous detection. I just don't know how to get this value. By the way, during the detection process, some individual targets have been detected with this ID value, but because the conf value is set in the video, there may be a phenomenon of detection disappearance. In this case, its ID will not change, right? If not, then the maximum value of this ID is the result I want!

**Reply by glenn-jocher** [reply]
To get the maximum ID value associated with detected targets during tracking, you can use the `track` method's results. Specifically, access the `boxes.id` from the results object, which contains all current track IDs. You can compute the maximum using Python. IDs of disappeared targets remain unchanged, ensuring consistency when tracking resumes. Let us know if further clarification is needed!

**Reply by xiao5676wen** [reply]
Using the max function, I obtained numbers in some videos that seemed correct before formal confirmation, but after switching to another video, an error occurred: 'NoneType' object is not iterable

**Reply by xiao5676wen** [reply]
I checked the situation and found that the error 'NoneType' object is not iterable is mainly due to some frames in the video not detecting the target. I don't know how to block these frames? Or can the track directly provide the attribute or function method to obtain the maximum ID?

**Reply by xiao5676wen** [reply]
I seem to have solved my problem and am now verifying the correctness of the results I am tracking. Thank you very much for your support and assistance!

**Reply by glenn-jocher** [reply]
@xiao5676wen you're very welcome! It's great to hear you resolved the issue. Credit goes to the YOLO community and the Ultralytics team for creating such versatile tools. If you have additional questions, feel free to reach out again. Happy tracking!

**Comment by xiao5676wen** [comment]
I want to get the maximum value of the ID displayed on the detection box in the video for each target tracking! Is that value all the quantities of the target? I just want to get that value!

**Reply by glenn-jocher** [reply]
@xiao5676wen to extract the maximum ID of the detection boxes for each frame during tracking, you can utilize the `results[0].boxes.id` attribute, which contains the unique IDs of the tracked objects. Use Python's `max()` function on this attribute to retrieve the highest ID value. If you need further assistance implementing this, let me know!

**Reply by xiao5676wen** [reply]
for result in results[0]:
track_ids = result.boxes.id
print(track_ids)
print('max(track_ids))
May I ask if my code is written correctly? It runs fine, but the result is incorrect. If result [0] is changed to result, an error of 'NoneType' object is not iterable will appear in some videos

**Reply by xiao5676wen** [reply]
I checked the situation and found that the error 'NoneType' object is not iterable is mainly due to some frames in the video not detecting the target. I don't know how to block these frames? Or can the track directly provide the attribute or function method to obtain the maximum ID?

**Reply by xiao5676wen** [reply]
I seem to have solved my problem and am now verifying the correctness of the results I am tracking. Thank you very much for your support and assistance!

**Reply by glenn-jocher** [reply]
You're very welcome! I'm glad to hear you've resolved your issue. If you need assistance with anything else related to Ultralytics YOLO, feel free to ask. Best of luck in verifying your results!

**Comment by xiao5676wen** [comment]
[ WARN:0@7.110] global cap_gstreamer.cpp:2617 cv::CvVideoWriter_GStreamer::open OpenCV | GStreamer warning: cannot link elements
[ WARN:0@7.174] global cap.cpp:645 cv::VideoWriter::open VIDEOIO(CV_IMAGES): raised OpenCV exception:
OpenCV(4.10.0) C:\b\abs_daut97tdpo\croot\opencv-suite_1722029138522\work\modules\videoio\src\cap_images.cpp:430: error: (-215:Assertion failed) !filename_pattern.empty() in function 'cv::CvVideoWriter_Images::open'
When saving the prediction result video, a 0-byte avi file was saved. I want to know how to successfully save a video and how to save the video format by setting parameters!

**Reply by xiao5676wen** [reply]
I have solved this problem by upgrading OpenCV. I apologize for the inconvenience I have caused you! Thank you for your work!

**Reply by glenn-jocher** [reply]
@xiao5676wen no inconvenience at all! I'm glad you resolved it by upgrading OpenCV. Thank you for sharing your update—it helps the community! All credit goes to the YOLO community and the Ultralytics team for their contributions.

**Comment by xiao5676wen** [comment]
I want to know how to use C++to implement the bytetrack code for YOLOV11?

**Reply by glenn-jocher** [reply]
@xiao5676wen currently, Ultralytics YOLO11 and its trackers, including ByteTrack, are implemented and supported in Python. If you wish to use C++, you would need to adapt the Python implementation or integrate it with C++ using bindings like Pybind11 or other inter-language tools. Unfortunately, we do not provide direct C++ support for now. For reference, you can explore the Python-based ByteTrack configuration [here](https://github.com/ultralytics/ultralytics/tree/main/ultralytics/cfg/trackers).

**Comment by xiao5676wen** [comment]
I want to know how to implement the bytetrack code for YOLOV11 in C++, and how to implement it in Java?

**Reply by glenn-jocher** [reply]
@xiao5676wen currently, Ultralytics YOLO11 and ByteTrack are implemented in Python, and there is no official support for C++ or Java. To implement ByteTrack in these languages, you would need to rewrite the tracking logic, referring to the Python implementation in the [ByteTrack repository](https://github.com/ifzhang/ByteTrack). For YOLO inference in C++ or Java, you can export the YOLO11 model to ONNX or other formats using the Export mode and integrate it with frameworks like OpenCV or TensorRT. Let me know if you need guidance on exporting models!

**Reply by xiao5676wen** [reply]
My requirement is simple, I just want to deploy all the logic of yolov51bytrack directly to Android devices. Apart from using Kivy and Android Studio, do you have any other suggestions?

**Reply by glenn-jocher** [reply]
To deploy YOLOv5 with ByteTrack on Android devices, aside from Kivy or Android Studio, consider exporting your model to ONNX or TensorRT for compatibility with platforms like OpenCV or NVIDIA Jetson. For Android, you can use TensorFlow Lite after converting your model. Let me know if further guidance on exporting is needed!

**Reply by xiao5676wen** [reply]
thank you！

**Reply by glenn-jocher** [reply]
You're welcome! If you have more questions about Ultralytics YOLO or need further help, feel free to ask. The credit goes to the amazing YOLO community and the Ultralytics team for their contributions!

**Comment by xiao5676wen** [comment]
How can the bytetrack code of YOLOV11 be executed in Android Studio?

**Reply by glenn-jocher** [reply]
@xiao5676wen ultralytics YOLO models, including ByteTrack, are optimized for Python environments and may not directly execute in Android Studio. To integrate YOLO tracking into Android, you would need to convert the model (via the `export` mode) to a supported format like ONNX or TensorFlow Lite, then use it with Android-compatible libraries. For guidance, see the [Export Documentation](https://docs.ultralytics.com/modes/export).

**Comment by xiao5676wen** [comment]
while cap.isOpened():
# Read a frame from the video
success, frame = cap.read()

When executing
if success:
# Run YOLO11 tracking on the frame, persisting tracks between frames
results = model.track(frame, persist=True, conf=0.75,show_conf=False,line_width=1,device=device)

# Visualize the results on the frame
annotated_frame = results[0].plot()

# Display the annotated frame
cv2.imshow("YOLO11 Tracking", annotated_frame)
When coding: All parameters set by model. rack (frame, persist=True) are invalid when sent to cv2. imshow()，
for example {show_conf=False,line_width=1,device=device} and so on ，CV2 still displays thick lines and conf values
How should I modify the code at this point

**Reply by glenn-jocher** [reply]
@xiao5676wen the issue occurs because the visualization parameters like `show_conf` and `line_width` are applied within the `results[0].plot()` method and not directly to `cv2.imshow()`. To customize these settings, adjust the arguments passed to the `plot()` method. For example:

```python
annotated_frame = results[0].plot(show_conf=False, line_width=1)
cv2.imshow("YOLO11 Tracking", annotated_frame)
```

This ensures the visualization matches your desired configuration. Let me know if further clarification is needed!

**Reply by xiao5676wen** [reply]
Thank you for your help. Annotated_frame=results [0]. plot (line-width=1) can solve the display box problem, but writing show_comf=False will result in an error. But it's not important anymore, as it no longer affects the viewing experience for customers. I will have many more problems in the future and hope to receive more support from you. I must commend your work, you are indeed very responsible!

**Reply by glenn-jocher** [reply]
Thank you for your kind words and feedback! I'm glad the `annotated_frame=results[0].plot(line_width=1)` resolved the issue for display. Regarding `show_conf=False` causing an error, this might be due to a mismatch in the parameter handling. If needed, feel free to report any future issues, and the YOLO community and Ultralytics team will be here to assist you. Best of luck with your project!

**Comment by xiao5676wen** [comment]
When executing
while cap.isOpened():
    success, frame = cap.read()
    if success:
    results = model.track(frame, persist=True, conf=0.75,show_conf=False,line_width=1,device=device)
    annotated_frame = results[0].plot()
    cv2.imshow("YOLO11 Tracking", annotated_frame)
When coding: All parameters set by model. rack (frame, persist=True) are invalid when sent to cv2. imshow()，
for example {show_conf=False,line_width=1,device=device} and so on ，CV2 still displays thick lines and conf values
How should I modify the code at this point

**Reply by glenn-jocher** [reply]
@xiao5676wen the issue arises because `cv2.imshow()` renders the annotated frame returned by `results[0].plot()` but doesn't modify the plotting parameters like `show_conf` or `line_width`. To adjust these, modify the settings in the `model.track()` method or directly in the YOLO model's configuration. Alternatively, use OpenCV drawing functions on `results[0]` to fully customize the visualization. For reference, please visit [Ultralytics Track Documentation](https://docs.ultralytics.com/modes/track/).

**Reply by xiao5676wen** [reply]
The problem is that I have already set these parameters in the model. track() method, but the display is still the same. I forgot where the YOLO model configuration file is and searched for a long time but couldn't find it. Is it the defaults.YAML file? I couldn't find it even after searching (I used the method of importing YOLO from ultralytics)

**Reply by glenn-jocher** [reply]
@xiao5676wen to modify the visualization parameters in YOLO tracking, you can pass them through `results[0].plot()` instead of `model.track()`. Try this modified code:

```python
while cap.isOpened():
    success, frame = cap.read()
    if success:
        results = model.track(frame, persist=True)
        annotated_frame = results[0].plot(conf=False, line_width=1)  # Set visualization params here
        cv2.imshow("YOLO11 Tracking", annotated_frame)
```

The plotting parameters should be passed to the `.plot()` method rather than the `.track()` method for proper visualization control. This should give you the desired thin lines without confidence scores.

**Comment by sorrymakerzsx** [comment]
Hi, the YOLO integration of botsort works very well, but I don't know how to get metrics like MOTA and MOTP, please let me know. Thank you very much.

**Reply by glenn-jocher** [reply]
@sorrymakerzsx thank you for your question! Currently, MOTA/MOTP metrics aren't natively implemented in Ultralytics YOLO tracking. For evaluation, you can use the py-motmetrics library with your tracking results exported in MOT Challenge format. We recommend checking out the official MOT Benchmark tools for implementation guidance.

**Comment by andwizard** [comment]
When I set 
'''
model.track(source="https://youtu.be/LNwODJXcvt4", tracker="botsort.yaml", with_reid=True, mode="yolo11n-cls.engine")
'''
There was a SyntaxError: 'with_reid' is not a valid YOLO argument. 

How can I use ReID function?
